# La thermistance: un capteur adapté ?

## Position du problème

En Sciences Physiques on a coutume pour étudier un phénomène de suivre des grandeurs mesurables.
Une grandeur mesurable très souvent utilisée est la **température**.
Comme on peut juger pratique de suivre automatiquement la température d'une réaction chimique, on peut se poser la question du choix du capteur à utiliser afin que celui-ci retranscrive le phénomène en jeu.


## La thermistance

<img align="right" width="100" height="100" src="https://capytale2.ac-paris.fr/web/sites/default/files/2022/05-20/23-42-30/thermistance.jpg">
La **thermistance**, ou **CTN**, est un composant électronique à deux pôles de connexion (dipôle) dont la valeur de résistance électrique interne (sa capacité à s'opposer au passage du courant électrique, donc au mouvement des électrons) va varier selon la température du milieu dans lequel on la place.

On schématise la CTN en utilisant la représentation suivante:

<img align="center" width="150" height="150" src="https://capytale2.ac-paris.fr/web/sites/default/files/2022/06-07/11-17-58/CTN.png">

La plupart du temps, ce composant est positionné dans un tube de protection ou une gaine thermoplastique dont l'extrémité est ouverte permettant le contact entre la CTN et le milieu dans lequel on souhaite suivre les variation de température:

<img align="center" width="250" height="250" src="https://capytale2.ac-paris.fr/web/sites/default/files/2022/06-07/12-16-30/CTN%20protection.png">

<img align="left" width="200" height="200" src="https://capytale2.ac-paris.fr/web/sites/default/files/2022/05-21/0-20-06/multimetre.png">Pour mesurer la valeur R de cette résistance en ohms, on utilise un multimètre en position ohm-mètre.

On branche l'ohm-mètre aux bornes de notre dipôle et on commence par utiliser le plus grand calibre disponible.

On descend alors dans l'échelle des calibres de façon à afficher sur l'écrane le plus grand nombre de chiffres significatifs.

On peut alors écrire la valeur mesurée $$R= xxx\;\Omega$$

## Matériel à disposition

Pour réaliser vos expériences, vous disposez du matériel suivant:
- ohm-mètre
- thermistance dans un "doigt de gant"
- fils de connexion
- bécher de 500 mL en Pyrex
- eau distillée
- glaçons
- agitateur magnétique chauffant réglable (en température et vitesse de rotation)
- thermomètre

## Schéma expérimental possible

On pourra se baser pour ce TP sur le montage suivant:
<img align="center" width="300" height="300" src="https://capytale2.ac-paris.fr/web/sites/default/files/2022/06-07/13-19-55/montage.png">

## Protocole expérimental

Précisez ci dessous le protocole expérimental que vous allez utiliser pour vérifier la capacité de la CTN à se comporter comme un vrai capteur de température.

**_votre protocole:_**

* <span style="color:green">On place dans un bécher de 500mL environ 250 mL d'eau distillée et une dizaine de petits glaçons.</span>
* <span style="color:green">On positionne dans le bécher un turbulent qui va permettre d'uniformiser la température du mélange.</span>
* <span style="color:green">On place le bécher sur l'agitateur chauffant en réglant le chauffage à T=60°C et agitation moyenne.</span>
* <span style="color:green">On place la sonde CTN (reliée à un ohm-mètre) et le thermomètre de façon à ce que ceux-ci ne touchent pas le fond du bécher mais se trouvent à mi profondeur.</span>
* <span style="color:green">On relève la température mesurée par le thermomètre et la valeur de R en ohms toutes minutes approximativement. Les valeurs sont placées dans un tableau.</span>
* <span style="color:green">On arrête nos relevés quand la température du liquide est à 60°C.</span>


## Analyse des résultats expérimentaux

> #### représentation des grandeurs acquises expérimentalement

Vous pouvez utiliser le code suivant pour tracer la courbe R=f(T).

In [ ]:
import matplotlib.pyplot as plt

R=[2.73e3,2.62e3,2.23e3,1.9e3,1.61e3,1.3e3,1.03e3,0.8e3,0.71e3,0.61e3,0.52e3,0.44e3,0.39e3,0.33e3,0.3e3,0.27e3]# on place les valeurs de R dans un premier tableau vide
T=[0.7,1.3,4.7,8.4,12.1,16.9,21.1,26,33,37,41,45.2,49,53,56.8,60.2]# on place les valeurs de T correspondantes dans un tableau vide


plt.plot(T,R,"b+") # La fonction plot(x,y) trace la courbe y=f(x) à partir des listes x et y .
plt.show() # affichage du graphe. 

Ce premier programme est incomplet, utilisez la boite à outil Python que vous avez complétée au cours des différents TP précédents pour permettre à ce programme d'afficher une échelle adaptée, d'afficher les titres des axes et le nom du graphe, de graduer celui-ci...

In [19]:
#Programme à compléter pour un affichage optimal

import matplotlib.pyplot as plt

R=[2.73e3,2.62e3,2.23e3,1.9e3,1.61e3,1.3e3,1.03e3,0.8e3,0.71e3,0.61e3,0.52e3,0.44e3,0.39e3,0.33e3,0.3e3,0.27e3]
T=[0.7,1.3,4.7,8.4,12.1,16.9,21.1,26,33,37,41,45.2,49,53,56.8,60.2]

plt.clf()#on efface le graphe précédent

plt.plot(T,R,"b+",label= ' R=f(T)')#on définit ce qui va être affiché et comment cela va s'afficher

plt.xlabel('T en °C')#on nomme les axes
plt.ylabel('R en ohms')

plt.xlim(0,70)#on définit les bornes des échelles
plt.ylim(0,3000)

plt.title('Résistance R de la CTN en fonction de la température T')#titre du graphique

plt.grid()# on gradue le graphe

plt.show() # affichage du graphe.

> #### Analyse de la courbe expérimentale :

**_votre réponse:_**
<span style="color:green">
On a pu remarquer que tant que des glaçons sont présents dans la bécher la température du mélange ne change pas,et donc la résistance de la CTN est stable également.
On a donc écarté les premiers couples de valeurs (R,T) pour le tracé de la courbe.
</span>    

<span style="color:green">On se rend compte que la résistance de la CTN baisse alors que la température augmente.
</span>

Potentiellement votre courbe est normalement modélisable.
Vous avez l'habitude, en sciences physiques, de travailler avec des modèles linéaires.

Informatiquement, on utilise un processus appelé **régression linéaire** qui permet, à partir des données expérimentales de trouver un modèle mathématique en accord avec la courbe.

Il faut pour cela utiliser une bibliothèque additionnelle, **"numpy"** et utiliser des tableaux vectoriels pour les variables mesurées.

In [ ]:
#on appelle la bibliothèque numpy en tant que np
import numpy as np

# forme vectorielle, on remplit les tableaux avec les valeurs expérimentales séparées par des ","
R = np.array([])
T = np.array([])

on va ensuite utiliser la fonction **polyfit** qui va calculer les coefficients de notre courbe de régression linéaire.

Mathématiquement cette fonction s'appuie sur l'utilisation de la méthode des moindres carrés pour un polynôme dont on fixera le degré à 1.

In [ ]:
coeff = np.polyfit(T,R,1) # la valeur 1 correspond au degré du polynôme de calcul pour la modélisation

Rmodel = coeff[0]*T+coeff[1] #permet de calculer les valeurs du modèle de R en fonction de T et des coefficients calculés de la courbe de régression

print('R={0:.2f}'.format(coeff[0]),'xT+{0:.2f}'.format(coeff[1])) # affiche l'équation de la courbe avec une précision à 2 décimale pour les coefficients

Regroupez ces lignes avec celles de votre premier programme afin d'afficher la courbe expérimentale et la courbe modélisée sur le même graphe.

In [28]:
#nouveau programme avec courbe de modélisation ici même
import matplotlib.pyplot as plt
import numpy as np

R = np.array([2.73e3,2.62e3,2.23e3,1.9e3,1.61e3,1.3e3,1.03e3,0.8e3,0.71e3,0.61e3,0.52e3,0.44e3,0.39e3,0.33e3,0.3e3,0.27e3])
T = np.array([0.7,1.3,4.7,8.4,12.1,16.9,21.1,26,33,37,41,45.2,49,53,56.8,60.2])
coeff = np.polyfit(T,R,1)

Rmodel = coeff[0]*T+coeff[1]

print('R = {0:.2f}'.format(coeff[0]),'* T + {0:.2f}'.format(coeff[1]))

plt.clf()#on efface le graphe précédent

plt.plot(T,R,"b+",label= ' R=f(T)')#on définit ce qui va être affiché et comment cela va s'afficher
plt.plot(T,Rmodel,"r-",label='modèle linéaire de R=f(T)')
plt.xlabel('T en °C')#on nomme les axes
plt.ylabel('R en ohms')

plt.xlim(0,70)#on définit les bornes des échelles
plt.ylim(0,3000)

plt.title('Résistance R de la CTN en fonction de la température T')#titre du graphique

plt.grid()# on gradue le graphe

plt.show() # affichage du graphe.

R = -39.04 * T + 2249.82


Quelle est l'équation de la courbe modélisée par votre programme ?

**_votre réponse :_**
<span style="color:green">la modélisation par régression linéaire donne une droite décroissante d'équation
R = -39.04 * T + 2249.82
</span>

A l'affichage de ces courbes, peut-on dire que la courbe modélisée par régression linéaire correspond aux valeurs expérimentales ? argumentez votre réponse.

**_votre réponse :_**

<span style="color:green">Il est évident que le modèle linéaire calculé ne correspond pas à la courbe réelle, l'une est droite alors que l'autre ressemble à une parabole inversée.</span>
<span style="color:green">    
On ne peut donc pas utiliser ce modèle pour faire une prédiction de correspondance entre la résistance et la température.</span>
<span style="color:green">    
Ce modèle ne correspond pas au fonctionnement de la CTN.</span>

> #### Conclusion :

concluez quand à la problématique initiale de ce TP en vous aidant de vos réponses précédentes.

**_votre réponse :_**
<span style="color:green">la CTN peut probablement être utilisée comme un capteur de température mais il faudrait avoir sa courbe d'étalonage en permanence disponible pour pouvoir faire correspondre la température avec la résistance.</span>

___

## Pour aller plus loin...

___

Lorsqu'on se trouve dans le cas d'une courbe qui ne correspond pas à un modèle linéaire, on peut songer à essayer de la modéliser par un polynôme.
Un polynôme peut s'écrire sous la forme:
$$ P(x)=p\left [ 0 \right ]\ast x^{deg}+p\left [ 1 \right ]\ast x^{deg-1}+...+p\left [ deg \right ]$$
On peut donc par exemple écrire les polynômes suivants:

polynôme de degré 2:
$$P(x)=p\left [ 0 \right ]\ast x^{2}+p\left [ 1 \right ]\ast x+p\left [ 2 \right ]$$
polynôme de degré 1:
$$P(x)=p\left [ 0 \right ]\ast x+p\left [ 1 \right ]$$
polynôme de degré 0:
$$P(x)=p\left [ 0 \right ]$$


Pour calculer un polynome de degré 2 à l'aide de **numpy**, on peut procéder en suivant l'exemple donné plus haut. Il faudra juste adapter un peu le code à écrire

In [ ]:
coeff = np.polyfit(T,R,2) #calcule les coeff du pôlynome de degré 2 à partir des données expérimentales
Rmodel = coeff[0]*(T**2)+coeff[1]*T+coeff[2] # calcule la fonction de modélisation 
print('R={0:.2f}'.format(coeff[0]),'*T^2 + {0:.2f}'.format(coeff[1]),'*T +{0:.2f}'.format(coeff[2])) # écrit l'équation du modèle

### votre job...

* recherchez quel est le polynôme qui va le mieux s'adapter à la courbe expérimentale de la CTN que vous avez utilisé
* affichez la courbe de modélisation en rouge par rapport à la courbe expérimentale
* déterminez quelle est l'équation de ce polynôme
* concluez à nouveau par rapport à la problématique initiale du TP, à savoir peut-on utiliser une CTN comme un capteur de température ?


In [34]:
#votre programme ici même
import numpy as np
import matplotlib.pyplot as plt

R = np.array([2.73e3,2.62e3,2.23e3,1.9e3,1.61e3,1.3e3,1.03e3,0.8e3,0.71e3,0.61e3,0.52e3,0.44e3,0.39e3,0.33e3,0.3e3,0.27e3])
T = np.array([0.7,1.3,4.7,8.4,12.1,16.9,21.1,26,33,37,41,45.2,49,53,56.8,60.2])
coeff = np.polyfit(T,R,4)

# on a essayé les polynômes suivants
# Rmodel = coeff[0]*(T**2)+coeff[1]*T+coeff[2] 
# Rmodel = coeff[0]*(T**3)+coeff[1]*(T**2)+coeff[2]*T+coeff[3]

# le modèle suivant correspond le mieux à la courbe expérimentale

Rmodel = coeff[0]*(T**4)+coeff[1]*(T**3)+coeff[2]*(T**2)+coeff[3]*T+coeff[4]

#affichage de l'équation
print('R = {0:.5f}'.format(coeff[0]),'* T^4 + {0:.3f}'.format(coeff[1]),'*T^3 + {0:.2f}'.format(coeff[2]), '* T^2 + {0:.2f}'.format(coeff[3]),'* T + {0:.2f}'.format(coeff[4]))
plt.clf()#on efface le graphe précédent

plt.plot(T,R,"b+",label= ' R=f(T)')#on définit ce qui va être affiché et comment cela va s'afficher
plt.plot(T,Rmodel,"r-",label='modèle linéaire de R=f(T)')
plt.xlabel('T en °C')#on nomme les axes
plt.ylabel('R en ohms')

plt.xlim(0,70)#on définit les bornes des échelles
plt.ylim(0,3000)

plt.title('Résistance R de la CTN en fonction de la température T')#titre du graphique

plt.grid()# on gradue le graphe

plt.show() # affichage du graphe.

R = 0.00014 * T^4 + -0.033 *T^3 + 2.99 * T^2 + -132.35 * T + 2805.68


**_vos réponses :_**

<span style="color:green">On voit d'après les simulations que le modèle qui correspond le mieux à la courbe réelle est un polynôme de degré 4 donc l'équation est:</span>

<span style="color:green">R = 0.00014 . T^4 + -0.033 . T^3 + 2.99 . T^2 + -132.35 . T + 2805.68</span>

<span style="color:green">On peut donc à priori à l'aide de ce modèle faire correspondre finement une température avec une une résistance mesurée pour la CTN étudiée.</span>

<span style="color:green">Dans ce cas la CTN est un capteur de température valable.</span>